tools/mkcomposefs.c

```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
 	return fwrite(buf, 1, count, file);
 }
 
+static void log_message_with_time( const char* msg )
+{
+	time_t t 		= time(NULL);
+  	struct tm tm 	= *localtime(&t);
+	fprintf(stdout, "%02d:%02d:%02d\t %s\n", 
+				tm.tm_hour, tm.tm_min, tm.tm_sec, msg);
+}
+
+struct ParallelCopyInfo
+{
+	struct lcfs_node_s		*node;
+	struct ParallelCopyInfo	*next;
+	char 					*parentPath;
+};
+
+static struct ParallelCopyInfo* new_parallel_copy_info(void)
+{
+	struct ParallelCopyInfo *ret = 
+		calloc(1, sizeof( struct ParallelCopyInfo));
+	ret->next 		= NULL;
+	ret->node 		= NULL;
+	ret->parentPath = NULL;
+
+	return ret;
+}
+
+static void cleanup_copy_node(struct ParallelCopyInfo *node)
+{
+	if(node) 
+	{
+		cleanup_copy_node(node->next);
+		free(node->parentPath);
+		free(node);
+	}	
+}
+
+static struct ParallelCopyInfo* construct_copy_data( struct lcfs_node_s* node, 
+													struct ParallelCopyInfo* current, 
+													char *path )
+{	
+	cleanup_free char 	*tmp_path 	= NULL;
+	const char 			*fname  	= lcfs_node_get_name(node);
+	if (fname) 
+	{
+		if (join_paths(&tmp_path, path, fname) < 0)
+			return NULL;
+
+		path = tmp_path;
+	}
+
+	if ( lcfs_node_dirp(node)) 
+	{
+		const size_t n_children = lcfs_node_get_n_children(node);
+		for (size_t i = 0; i < n_children; i++)
+		{
+			current = construct_copy_data(lcfs_node_get_child(node, i), 
+											current, path);
+		} 
+	}
+	else if ((lcfs_node_get_mode(node) & S_IFMT) == S_IFREG &&
+		   		lcfs_node_get_content(node) == NULL 		&&
+		   		lcfs_node_get_payload(node) != NULL) 
+	{
+		current->node 		= node;
+		current->parentPath = strdup(path);
```

Comment:
check strdup return value (and in other strdup calls)

Reply:
Corrected